### DISCUSSION TITLE
Fine Tuning Strength

### DISCUSSION THREAD

**Discussion by nealmick** [nealmick/bb#32] [Announcements]
When training a model games are batched in a group and computed one batch at a time.  This is great for speeding up computation and can also help generalize based on larger groups.  Model retraining has been previously offered only with default learning rates specified by the optimizer.  This causes some issues, Originally the model was trained computing (lets just say) batch size 128 games at a time, and now when we want to retrain on game, we only have one game, not 128 to batch together.  What’s the solution? Set batch size to 1 when retraining.  Great, this seemed to make it work initially however after multiple retrains, the models seem to forget original patterns and seems to be less accurate.  I think what effectively happens is the one game thats loaded with batch size 1, to the model has a weight of what 128 previous training games were….. Okay soo if we only have 1 game how can we make batch size 1 execute with the same level importance to model?  Variable Learning Rates.  Today there is a new feature on the live server that allows for a percentage of the original learning rate to be used when retraining.  Currently when you specify an optimizer on the training dashboard, that optimizer come’s along with TensorFlow default parameters, including learning rate.  Now on the live server there is a new slider next to the retrain button.   The slider ranges from 1% to 200%, when retrain is clicked the learning rate % is sent to the server, once the server gets request the model settings are loaded including the default learning rate.  Once we have the default learning rate, we then create a custom optimizer with a percentage of the old learning rate.  The model is then trained with a single epoch, and a batch size of 1, using the new custom optimizer and adjusted learning rate.  This feature is currently only available under 10$ a month plan on the live server.   Let me know if any has any questions!

**Comment by nealmick** [comment]
![Screen Shot 2023-11-17 at 1 41 43 PM](https://github.com/nealmick/bb/assets/33704976/eb66fc23-045c-4717-8da7-6eb4b147b9c3)
